### PR TITLE
[REVIEW] Fix ORC reader issue with decimal type

### LINKED ISSUE #5892: [BUG] Incorrect values with read_orc
**Describe the bug**
Getting a difference in values when reading a file with `cudf.read_orc` vs `pd.read_orc / pyarrow.orc`

**Steps/Code to reproduce bug**
```python
import cudf
df = cudf.read_orc("orc_mismatch.orc")
df._col8.value_counts()

1.166878e-02     99677
1.166878e+16         6
4.829986e-320        1
Name: _col8, dtype: int32

df._col8.dtype
dtype('float64')
```

**Expected behavior**
```python
import pandas as pd
df = pd.read_orc("orc_mismatch.orc")
df._col8.value_counts()

0.011668778935185185    99684
Name: _col8, dtype: int64

df._col8.dtype
dtype('O')
```

**Environment overview (please complete the following information)**
 - Environment location: Baremetal 
 - Method of cuDF install: conda

**Environment details**
Nightly from Aug7

**Additional context**
Link to data: [orc_mismatch.orc.zip](https://github.com/rapidsai/cudf/files/5044005/orc_mismatch.orc.zip)

Original type `_col8` as interpreted by PyArrow  is `decimal`

```python
from pyarrow import orc

a = orc.ORCFile("orc_mismatch.orc")
a.schema

_col0: int32
_col8: decimal(38, 18)
```
<details>
<summary>Here are the statistics</summary>
<pre>
$ orc-statistics orc_mismatch.orc

*** Column 0 ***
Column has 99684 values and has null value: no

*** Column 1 ***
Data type: Integer
Values: 99684
Has null: no
Minimum: 1
Maximum: 7326
Sum: 230857068

*** Column 2 ***
Data type: Decimal
Values: 99684
Has null: no
Minimum: 0.011668778935185185
Maximum: 0.011668778935185185
Sum: 1163.19055937499998154

File orc_mismatch.orc has 1 stripes
*** Stripe 0 ***

--- Column 0 ---
Column has 99684 values and has null value: no

--- Column 1 ---
Data type: Integer
Values: 99684
Has null: no
Minimum: 1
Maximum: 7326
Sum: 230857068

--- Column 2 ---
Data type: Decimal
Values: 99684
Has null: no
Minimum: 0.011668778935185185
Maximum: 0.011668778935185185
Sum: 1163.19055937499998154
</pre>
</details>

cc: @mikechenfu Thanks for pointing out the issue! 😃

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Number of values being read was't considering last update while breaking, which caused the subsequent calls behave oddly, leading to improper access of memory segments.

closes #5892

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by rgsl888prabhu** [comment]
@harrism should we push this 0.16 as this is priority-0 and and change is just one line.

**Comment by kkraus14** [comment]
@rgsl888prabhu is there a small test file we could add to cover the error case here?

**Comment by rgsl888prabhu** [comment]
> @rgsl888prabhu is there a small test file we could add to cover the error case here?

It is a 64KB file with which we can reproduce, would it be fine to add it ?

**Comment by vuule** [comment]
> It is a 64KB file with which we can reproduce, would it be fine to add it ?

Is it possible to make a smaller repro?

**Comment by rgsl888prabhu** [comment]
> > It is a 64KB file with which we can reproduce, would it be fine to add it ?
> 
> Is it possible to make a smaller repro?
> Edit: 64K not that large, but it looks like this issue should be reproducible with a much smaller file.

I am trying to create smaller file to test it.

**Review by harrism — APPROVED** [review]

**Comment by OlivierNV** [comment]
Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?

**Comment by rgsl888prabhu** [comment]
> Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?
> [Edit] I think we might need to limit the max # of scale values in RLE stream1 such that it can't overflow when decoding the secondary base128 stream, eg: 512-limit for max run = NTHREADS/2 at line 1551 (might exceed the max internal buffer size, and I don't think that's a case where overflow can be properly handled because of cross-stream dependency).
> Long-term, this kernel should really be rewritten with 512-threads with separate queues for primary & secondary stream as it would simplify things a lot and likely better for perf when index is available (it's a big change though).

```
if (t == 0) {
    uint32_t maxpos  = min(bs->len, bs->pos + (BYTESTREAM_BFRSZ - 8u));
    uint32_t lastpos = bs->pos;
    uint32_t n;
    for (n = 0; n < numvals; n++) {
      uint32_t pos                  = lastpos;
      *(volatile int32_t *)&vals[n] = lastpos;
      pos += varint_length<uint4>(bs, pos);
      if (pos > maxpos) break;
      lastpos = pos;
    }
    scratch->num_vals = n;
    bytestream_flush_bytes(bs, lastpos - bs->pos);
  }
```

This issue occurs when numvals=1024, so for n=1023,  pos > max_pos and it would break. But at n=1023, we have already update the value val[n] with last_pos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.

**Comment by rgsl888prabhu** [comment]
> Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?
> [Edit] I think we might need to limit the max # of scale values in RLE stream1 such that it can't overflow when decoding the secondary base128 stream, eg: 512-limit for max run = NTHREADS/2 at line 1551 (might exceed the max internal buffer size, and I don't think that's a case where overflow can be properly handled because of cross-stream dependency).
> Long-term, this kernel should really be rewritten with 512-threads with separate queues for primary & secondary stream as it would simplify things a lot and likely better for perf when index is available (it's a big change though).

@OlivierNV  And as you suggested, changing the `max run = NTHREADS/2 ` also fixes the issue, but that creates issue in other scenarios as you have already mentioned.

**Review by nvdbaranec — APPROVED** [review]

**Comment by OlivierNV** [comment]
> This issue occurs when numvals=1024, so for n=1023, pos > max_pos and it would break. But at n=1023, we have already updated the value val[n] with lastpos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.

It's storing the last offset but not using it (which is likely problem since that same location is initially used to store the scale).
What about moving the line `*(volatile int32_t *)&vals[n] = lastpos;` after `if (pos > maxpos) break;` ?

**Comment by rgsl888prabhu** [comment]
> > Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?
> > [Edit] I think we might need to limit the max # of scale values in RLE stream1 such that it can't overflow when decoding the secondary base128 stream, eg: 512-limit for max run = NTHREADS/2 at line 1551 (might exceed the max internal buffer size, and I don't think that's a case where overflow can be properly handled because of cross-stream dependency).
> > Long-term, this kernel should really be rewritten with 512-threads with separate queues for primary & secondary stream as it would simplify things a lot and likely better for perf when index is available (it's a big change though).
> 
> @OlivierNV
> 
> ```
> if (t == 0) {
>     uint32_t maxpos  = min(bs->len, bs->pos + (BYTESTREAM_BFRSZ - 8u));
>     uint32_t lastpos = bs->pos;
>     uint32_t n;
>     for (n = 0; n < numvals; n++) {
>       uint32_t pos                  = lastpos;
>       *(volatile int32_t *)&vals[n] = lastpos;
>       pos += varint_length<uint4>(bs, pos);
>       if (pos > maxpos) break;
>       lastpos = pos;
>     }
>     scratch->num_vals = n;
>     bytestream_flush_bytes(bs, lastpos - bs->pos);
>   }
> ```
> 
> This issue occurs when numvals=1024, so for n=1023, pos > max_pos and it would break. But at n=1023, we have already updated the value val[n] with lastpos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.



> > This issue occurs when numvals=1024, so for n=1023, pos > max_pos and it would break. But at n=1023, we have already updated the value val[n] with lastpos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.
> 
> It's storing the last offset but not using it (which is likely problem since that same location is initially used to store the scale).
> What about moving the line `*(volatile int32_t *)&vals[n] = lastpos;` after `if (pos > maxpos) break;` ?

It still causes same issue if I move the line after break.

**Comment by OlivierNV** [comment]
Yeah, it's not handling the buffering of scale values for decimals, and the buffer capacity is only sized for 512x64bit values, so will not handle 512 worst-case 128b decimals in one-shot consumption.
Along with moving that line after the break, it's missing code at around line 1811 (with the comment "buffer secondary stream values") to deal with buffering the leftover decimal scale values for the next pass (only done for BOOLEAN and TIMESTAMP types). The leftover values need to be stored back to the beginning of the buffer and buffered_count adjusted accordingly.
Also, it needs some added logic around line 1650 to deal with the buffered values, since in the case of decimals, we're really buffering data from the primary data stream (not a trivial change).
As a short-term WAR, we could limit the initial value of max_vals to 512 for DECIMAL type (would work here), but it's not a good long-term solution since even 512 decimal values could overflow the stream buffer (worst case is (128+6)/7 bytes per value)

**Comment by OlivierNV** [comment]
Or alternatively to the cross-iteration buffering, a simpler solution might be to add a loop within Decode_Decimals and refill the buffer on the fly (likely less ugly than the buffering).

**Comment by rgsl888prabhu** [comment]
Don't merge this, still under progress !!!!!!!!!!!!!!!!!!!!!!!!!

**Review by kkraus14 — CHANGES_REQUESTED** [review]
blocking merging

**Comment by rgsl888prabhu** [comment]
github is experiencing issues, and that's why my commit is not yet being shown in this PR.

File: cpp/src/io/orc/stripe_data.cu
Comment (OlivierNV) [inline cpp/src/io/orc/stripe_data.cu:1103]: Needs a `__syncthreads()` between `bytestream_fill()` and `if (t == 0) {` (race condition with `bs->fill_count`)

**Review by OlivierNV — APPROVED** [review]
LGTM

**Comment by rgsl888prabhu** [comment]
Please select `Hide whitespace changes` option as normal diff shows lot of unrelated things.

**Comment by kkraus14** [comment]
@rgsl888prabhu did you forget to check in the test files?

**Comment by vuule** [comment]
> @rgsl888prabhu did you forget to check in the test files?

I see the Python test file change

**Comment by kkraus14** [comment]
The Python test file refers to: `TestOrcFile.decimal.same.values.orc` and `TestOrcFile.decimal.multiple.values.orc` which aren't checked into CI as far as I could tell

**Review by kkraus14 — APPROVED** [review]

**Comment by rgsl888prabhu** [comment]
> @rgsl888prabhu did you forget to check in the test files?

@vuule He meant orc format files for tests.

@kkraus14  I just uploaded them.

**Comment by kkraus14** [comment]
@vuule do you want to review before we mark this as ready to merge?

**Comment by vuule** [comment]
@rgsl888prabhu Do you think we should check for perf regression before merging this? Looks unlikely to me, but I don't know the context well.

**Review by vuule — APPROVED** [review]
Good stuff. Thanks for finding smaller repro files.

**Comment by rgsl888prabhu** [comment]
> Good stuff. Thanks for finding smaller repro files.

Thanks to @OlivierNV.

**Comment by rgsl888prabhu** [comment]
> @rgsl888prabhu Do you think we should check for perf regression before merging this? Looks unlikely to me, but I don't know the context well.

Will try it and let you know.

**Comment by rgsl888prabhu** [comment]
Tried with a ORC dataset with just one column with `440000000`, orc file size is 250 MB.

There is not much change in spped.
older
```
In [2]: %timeit cudf.read_orc("/home/rgsl888/orc_8GB_250MB.orc")
1.7 s ± 6.05 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

```
In [2]: %timeit cudf.read_orc("/home/rgsl888/orc_8GB_250MB.orc")
1.64 s ± 260 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**
